test/controllers/APIControllerSpec.scala

```diff
@@ -0,0 +1,163 @@
+package controllers
+
+import akka.stream.Materializer
+import akka.util.ByteString
+import io.circe.{Json, JsonObject}
+import io.circe.syntax.EncoderOps
+import models.ProductFromJson
+import org.scalatestplus.play._
+import org.scalatestplus.play.guice.GuiceOneAppPerTest
+import play.api.Play.materializer
+import play.api.test.Helpers._
+import play.api.test.{FakeHeaders, FakeRequest, Injecting}
+import play.api.mvc._
+
+import java.util.UUID
+import scala.concurrent.{ExecutionContext, Future}
+import scala.language.postfixOps
+import play.api.db.Databases
+import play.api.inject.guice.GuiceApplicationBuilder
+import play.api.libs.streams.Accumulator
+
+class APIControllerSpec
+    extends PlaySpec
+    with Results
+    with GuiceOneAppPerTest
+    with Injecting {
+
+  def mockApp = new GuiceApplicationBuilder().build()
+  val mtrlzr = mockApp.injector.instanceOf[Materializer]
+
+  val database = Databases(
+    driver = "org.h2.Driver",
+    url = "jdbc:h2:mem:play"
+  )
+
+  "postProduct method" should {
+    "return successful when adding a product with all needed values" in {
+      val controller = inject[APIController]
+      val product = ProductFromJson(Some("crepe"), Some(5), Some(4.99)).asJson
+      val fakeRequest = FakeRequest(
+        POST,
+        controllers.routes.APIController.postProduct().url
+      ).withBody(product)
+        .withHeaders(("Content-Type" -> "application/json"))
+
+      val result: Accumulator[ByteString, Result] = controller
+        .postProduct()
+        .apply(
+          fakeRequest
+        )
+      val runResult: Future[Result] = result.run()(mtrlzr)
+
+      status(runResult) mustBe 200
+    }
```

Comment:
The problem is that the framework expects the same types to be used when you call the relevant methods.

 Specifically, it seems that the `apply()` on line 48 is taking on a `ProductFromJson` (which is what you are passing to your controller) and your tests expect that. One way to fix this is by making the following changes:

```suggestion
    "return successful when adding a product with all needed values" in {
      val controller = inject[APIController]
      val product = ProductFromJson(Some("crepe"), Some(5), Some(4.99))
      val fakeRequest = FakeRequest(
        POST,
        controllers.routes.APIController.postProduct().url
      ).withBody(product)
        .withHeaders(("Content-Type" -> "application/json"))

      val result: Future[Result] = controller
        .postProduct()
        .apply(
          fakeRequest
        )

      status(result) mustBe 200
    }
```
that way the apply for the `postProduct()` is taking on the same type as the one expected from your controller

Reply:
Thank you for showing me this, it helped immensely. I tried doing this earlier, but I was missing a couple of things that stopped me from doing it successfully: asJson on line 48 for example.